### PR TITLE
[SPARK-43944][CONNECT][PYTHON] Add string functions to Scala and Python - part 2

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
Add following functions:

- replace
- split_part
- substr
- parse_url
- printf
- url_decode
- url_encode
- position
- endswith
- startswith

to:

- Scala API
- Python API
- Spark Connect Scala Client
- Spark Connect Python Client

### Why are the changes needed?
for parity

### Does this PR introduce _any_ user-facing change?
Yes, new functions.

### How was this patch tested?
- Add New UT.
- Pass GA.

### CODE REVIEW

**Comment by panbingkun** [comment]
cc @zhengruifeng

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -3815,6 +3815,180 @@ object functions {
...
+  def position(substr: Column, str: Column): Column = withExpr {
+    new StringLocate(substr.expr, str.expr)
+  }
+
+  /**
+   * Returns a boolean. The value is True if str ends with suffix.
+   * Returns NULL if either input expression is NULL. Otherwise, returns False.
+   * Both str or suffix must be of STRING type.
+   *
+   * @note
```
Comment (zhengruifeng) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:3966]: please update the `@note` here

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 2]
```diff
@@ -3815,6 +3815,180 @@ object functions {
...
+  def endswith(str: Column, suffix: Column): Column = withExpr {
+    EndsWith(str.expr, suffix.expr)
+  }
+
+  /**
+   * Returns a boolean. The value is True if str starts with prefix.
+   * Returns NULL if either input expression is NULL. Otherwise, returns False.
+   * Both str or prefix must be of STRING type.
+   *
+   * @note
```
Comment (zhengruifeng) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:3981]: ditto

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
also cc @beliefer

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:3966]: +1

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala [hunk 3]
```diff
@@ -2344,7 +2344,63 @@ class PlanGenerationTestSuite
...
   functionTest("to_number") {
-    fn.to_char(fn.col("g"), lit("$99.99"))
+    fn.to_number(fn.col("g"), lit("$99.99"))
+  }
+
+  functionTest("replace") {
+    fn.replace(fn.col("g"), fn.col("g"))
+  }
+
+  functionTest("replace with replace") {
```
Comment (beliefer) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala:2354]: replace with specified string

**Comment by panbingkun** [comment]
This is all done.

**Comment by zhengruifeng** [comment]
thanks, merged to master

**Pull request closed without merging**